app/controllers/v3/roles_controller.rb

```diff
@@ -65,7 +65,8 @@ def destroy
     resource_not_found!(:role) unless role
 
     if role.for_space?
-      org_guid = Space.find(guid: role.space_guid).organization.guid
+      org_guid = Organization.join(:spaces, organization_id: :id).select(:organizations__guid).where(spaces__guid: role.space_guid)
```

Comment:
is this really faster?  I feel like it may be easier to call
```
org_guid = Space.find(guid: role.space_guid).organization_guid
```

as that will not query the DB to pull up the entire organization

Reply:
hmm, can we not just call `role.organization_guid`?

Reply:
I ran a test with three different versions:

1. `Space.find(guid: role.space_guid).organization.guid`

```
(app/models/runtime/role.rb:110:in `space_guid') SELECT * FROM `spaces` WHERE `id` = 1
(app/controllers/v3/roles_controller.rb:68:in `destroy') SELECT * FROM `spaces` WHERE (`guid` = 'big-space') LIMIT 1
(app/controllers/v3/roles_controller.rb:68:in `destroy') SELECT * FROM `organizations` WHERE `id` = 2
(lib/cloud_controller/membership.rb:22:in `has_any_roles?') SELECT 1 AS `one` FROM (...) AS `t1` WHERE ((`type` IN ('space_manager')) AND (`user_id` = 3) AND (`space_id` IN (SELECT `id` FROM `spaces` WHERE (`guid` = 'big-space')))) LIMIT 1
(lib/cloud_controller/membership.rb:27:in `has_any_roles?') SELECT 1 AS `one` FROM (...) AS `t1` WHERE ((`type` IN ('organization_manager')) AND (`user_id` = 3) AND (`organization_id` IN (SELECT `id` FROM `organizations` WHERE (`guid` = 'big-org')))) LIMIT 1
```

There are two queries to get the org guid:
```
SELECT * FROM `spaces` WHERE (`guid` = 'big-space') LIMIT 1
SELECT * FROM `organizations` WHERE `id` = 2
```

2. `Space.find(guid: role.space_guid).organization_guid`

```
(app/models/runtime/role.rb:110:in `space_guid') SELECT * FROM `spaces` WHERE `id` = 1
(app/controllers/v3/roles_controller.rb:68:in `destroy') SELECT * FROM `spaces` WHERE (`guid` = 'big-space') LIMIT 1
(lib/sequel_plugins/vcap_relations.rb:124:in `block in define_guid_reader') SELECT * FROM `organizations` WHERE `id` = 2
(lib/cloud_controller/membership.rb:22:in `has_any_roles?') SELECT 1 AS `one` FROM (...) AS `t1` WHERE ((`type` IN ('space_manager')) AND (`user_id` = 3) AND (`space_id` IN (SELECT `id` FROM `spaces` WHERE (`guid` = 'big-space')))) LIMIT 1
(lib/cloud_controller/membership.rb:27:in `has_any_roles?') SELECT 1 AS `one` FROM (...) AS `t1` WHERE ((`type` IN ('organization_manager')) AND (`user_id` = 3) AND (`organization_id` IN (SELECT `id` FROM `organizations` WHERE (`guid` = 'big-org')))) LIMIT 1
```

The exact same two queries are executed, but this time the second query comes from the `VcapRelations` sequel plugin as there is no `organization_guid` column in the `spaces` table. I wasn't aware of this magic...

3. `Organization.join(:spaces, organization_id: :id).select(:organizations__guid).where(spaces__guid: role.space_guid)`

```
(app/models/runtime/role.rb:110:in `space_guid') SELECT * FROM `spaces` WHERE `id` = 1
(lib/cloud_controller/membership.rb:22:in `has_any_roles?') SELECT 1 AS `one` FROM (...) AS `t1` WHERE ((`type` IN ('space_manager')) AND (`user_id` = 3) AND (`space_id` IN (SELECT `id` FROM `spaces` WHERE (`guid` = 'big-space')))) LIMIT 1
(lib/cloud_controller/membership.rb:27:in `has_any_roles?') SELECT 1 AS `one` FROM (...) AS `t1` WHERE ((`type` IN ('organization_manager')) AND (`user_id` = 3) AND (`organization_id` IN (SELECT `id` FROM `organizations` WHERE (`guid` IN (SELECT `organizations`.`guid` FROM `organizations` INNER JOIN `spaces` ON (`spaces`.`organization_id` = `organizations`.`id`) WHERE (`spaces`.`guid` = 'big-space')))))) LIMIT 1
```

Now, there is a single query to get the org guid (`SELECT 'organizations'.'guid' FROM 'organizations' INNER JOIN 'spaces' ON ('spaces'.'organization_id' = 'organizations'.'id') WHERE ('spaces'.'guid' = 'big-space')`) and this query is even inlined into the roles check and thus sometimes not even executed (if the user is authorized by space role). So for this specific query, I think that Sam's proposal is the most efficient one.

@samcolson4: I would prefer if the variable name is changed from `org_guid` to `org_guid_query` to be consistent with other places.

P.S. We should think about enhancing `VCAP::CloudController::Permissions` and `VCAP::CloudController::Membership` to (alternatively) deal with space and org ids instead of guids. Now we often have an id, get the guid for this id and finally get the id for this guid...

Reply:
What actually would work is a combination of @sethboyles' ideas:
```
unauthorized! unless permission_queryer.can_update_active_space?(role.space_guid, role.space.organization_guid)
```

`role.space_guid` loads the space, and then `role.space.organization_guid` loads the org.

This would mean 4 queries (i.e. one less than no. 1 and 2 above / one more than no. 3). So it would still be an improvement and a very readable solution...

Reply:
ah, I wasn't aware of the magic with `organization_guid`, either. 

@philippthun I think your combination solution with the additional commit looks good to me.  

Just curious--did you all notice a performance issue with this endpoint?  I would have thought that being a DELETE endpoint it would not be called all too often and cause too many issues.

Reply:
I'm happy to adopt that change for the sake readability. I've cherry-picked Philipp's [commit](https://github.com/sap-contributions/cloud_controller_ng/commit/aadce573387cd6c8d5aec3fba67d8d12ae59af25) into the branch, and have switched to `role.organization_guid` for space roles.

As a result of that I've tweaked [app/decorators/include_role_organization_decorator.rb](https://github.com/cloudfoundry/cloud_controller_ng/pull/2971/files#diff-285ddf85dfb392e2996c3040308af7407fac13b0077f84c6efef4b9fc8d06c8e) so as to avoid retrieving orgs for space roles when hitting `/v3/roles?include=organization` (previously this retrieved the org for any role for whom `organization_guid` wasn't nil).

Reply:
@sethboyles I threw a visualization together for one of our larger foundations giving the _sum_ of response times for different endpoints and request methods, excluding what I reckoned were most those whose slowness could plausibly be attributed to some external source (i.e. a service broker or Diego). On this measure the CREATE/DELETE endpoints for roles were both in the top ten, and since they concerned the same controller I decided to tackle them together.